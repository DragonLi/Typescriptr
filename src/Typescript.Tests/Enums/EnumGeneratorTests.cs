using System;
using Assent;
using Typescriptr;
using Xunit;

namespace Typescript.Tests.Enums
{
    public class EnumGeneratorTests
    {
        class TypeWithEnum
        {
            public enum EnumType
            {
                FirstEnum,
                SecondEnum,
                ThirdEnum
            }

            public EnumType AnEnum { get; set; }
        }

        [Fact]
        public void Generator_TypeWithEnum_GeneratesSuccessfully()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithEnum)});

            var result = string.Join($"{Environment.NewLine}---{Environment.NewLine}", generated.Types,
                generated.Enums);

            this.Assent(result);
        }
        
        class TypeWithNullableEnum
        {
            public enum EnumType
            {
                FirstEnum,
                SecondEnum,
                ThirdEnum,
            }

            public EnumType? NullableEnumProp { get; set; }
        }

        [Fact]
        public void Generator_TypeWithNullableEnum_GeneratesSuccessfully()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithNullableEnum)});

            this.Assent(generated.JoinTypesAndEnums());
        }


        class TypeWithValuedEnum
        {
            public enum EnumType
            {
                First = 1,
                Second = 2,
                Third = 3,
            }
            
            public EnumType EnumWithValueProp { get; set; }
        }

        [Fact]
        public void Generator_TypeWithEnumAndEnumValueFormatter_RendersValues()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithValuedEnum)});

            this.Assent(generated.JoinTypesAndEnums());
        }

        
        class TypeOneWithEnum
        {
            public SharedEnumType AnEnum { get; set; }
        }
        class TypeTwoWithEnum
        {
            public SharedEnumType AnEnum { get; set; }
        }

        public enum SharedEnumType
        {
            FirstEnum,
            SecondEnum,
            ThirdEnum
        }
        
        [Fact]
        public void Generator_TypesWithSharedEnum_ShouldOnlyGenerateTheEnumOnce()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeOneWithEnum), typeof(TypeTwoWithEnum)});

            this.Assent(generated.JoinTypesAndEnums());
        }
    }
}