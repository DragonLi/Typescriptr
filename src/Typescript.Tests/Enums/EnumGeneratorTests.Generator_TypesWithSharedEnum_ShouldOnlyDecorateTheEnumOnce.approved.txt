declare namespace Api {
  // Source: Typescript.Tests.Enums.EnumGeneratorTests+TypeTwoWithEnum
  interface TypeTwoWithEnum {
    anEnum: 'FirstEnum' | 'SecondEnum' | 'ThirdEnum';
  }
  // Source: Typescript.Tests.Enums.EnumGeneratorTests+TypeOneWithEnum
  interface TypeOneWithEnum {
    anEnum: 'FirstEnum' | 'SecondEnum' | 'ThirdEnum';
  }
}

---
// Source: Typescript.Tests.Enums.EnumGeneratorTests+SharedEnumType
enum SharedEnumType {
  FirstEnum = 'FirstEnum',
  SecondEnum = 'SecondEnum',
  ThirdEnum = 'ThirdEnum',
}
